File: src/main/java/edu/neu/coe/info6205/sort/simple/QuickSort_3way.java
package edu.neu.coe.info6205.sort.simple;

import edu.neu.coe.info6205.sort.BaseHelper;
import edu.neu.coe.info6205.sort.Helper;
import edu.neu.coe.info6205.sort.InstrumentedHelper;
import edu.neu.coe.info6205.util.Config;

import java.util.ArrayList;
import java.util.List;

public class QuickSort_3way<X extends Comparable<X>> extends QuickSort<X> {

    public static final String DESCRIPTION = "QuickSort 3 way";

    /**
     * Constructor for QuickSort_3way
     *
     * @param helper an explicit instance of Helper to be used.
     */
    public QuickSort_3way(Helper<X> helper) {
        super(helper);
        setPartitioner(createPartitioner());
    }

    public QuickSort_3way() {
        this(new BaseHelper<>(DESCRIPTION));
    }

    /**
     * Constructor for QuickSort_3way
     *
     * @param N      the number elements we expect to sort.
     * @param config the configuration.
     */
    public QuickSort_3way(int N, Config config) {
        super(DESCRIPTION, N, config);
        setPartitioner(createPartitioner());
    }

    /**
     * Constructor for QuickSort_3way which always uses an instrumented helper with a specific seed.
     * <p>
     * NOTE used by unit tests.
     *
     * @param N    the number of elements to be sorted.
     * @param seed the seed for the random number generator.
     */
    public QuickSort_3way(int N, long seed, Config config) {
        this(new InstrumentedHelper<>(DESCRIPTION, N, config));
    }

    public Partitioner<X> createPartitioner() {
        return new Partitioner_3Way(getHelper());
    }

    class Partitioner_3Way implements Partitioner<X> {

        /**
         * Method to partition the given partition into smaller partitions.
         *
         * @param partition the partition to divide up.
         * @return an array of partitions, whose length depends on the sorting method being used.
         */
        public List<Partition<X>> partition(Partition<X> partition) {
            // CONSIDER merge with Partitioner_DualPivot
            X[] xs = partition.xs;
            int lt = partition.from;
            int gt = partition.to - 1;
            helper.swapConditional(xs, lt, gt);
            X v = xs[lt];
            int i = lt + 1;
            // NOTE: we are trying to avoid checking on instrumented for every time in the inner loop for performance reasons (probably a silly idea).
            // NOTE: if we were using Scala, it would be easy to set up a comparer function and a swapper function. With java, it's possible but much messier.
            if (helper.instrumented())
                while (i <= gt) {
                    int cmp = helper.compare(xs[i], v);
                    if (cmp < 0) helper.swap(xs, lt++, i++);
                    else if (cmp > 0) helper.swap(xs, i, gt--);
                    else i++;
                }
            else
                while (i <= gt) {
                    int cmp = xs[i].compareTo(v);
                    if (cmp < 0) swap(xs, lt++, i++);
                    else if (cmp > 0) swap(xs, i, gt--);
                    else i++;
                }

            List<Partition<X>> partitions = new ArrayList<>();
            partitions.add(new Partition<>(xs, partition.from, lt));
            partitions.add(new Partition<>(xs, gt + 1, partition.to));
            return partitions;
        }

        public Partitioner_3Way(Helper<X> helper) {
            this.helper = helper;
        }

        private void swap(X[] ys, int i, int j) {
            if (helper != null) helper.swap(ys, i, j);
            else {
                X temp = ys[i];
                ys[i] = ys[j];
                ys[j] = temp;
            }
        }

        private final Helper<X> helper;
    }
}

File: Python/src/sort/simple/quick_sort_3way.py
from dataclasses import dataclass
from typing import List

from sort.simple.helper import Helper
from sort.simple.sort import Sort
from util.generic_type import C


class QuickSort3way(Sort[C]):

    def __init__(self, helper: Helper[C] = None):
        if helper is None:
            helper = Helper[C]("3-way QuickSort")
        self.helper = helper

    @dataclass
    class Partition:
        lt: int
        gt: int

    def _sort(self, xs: List[C], _from: int, _to: int) -> None:
        lo = _from
        hi = _to - 1
        if hi <= lo:
            return
        partition = self.partition(xs, lo, hi)
        self._sort(xs, lo, partition.lt - 1)
        self._sort(xs, partition.gt + 1, hi)

    def partition(self, xs: List[C], lo: int, hi: int) -> Partition:
        lt = lo
        gt = hi
        if self.helper.less(xs[hi], xs[lo]):
            self.swap(xs, lo, hi)
        v = xs[lo]
        i = lo + 1
        while i <= gt:
            cmp = self.helper.compare(xs[i], v)
            if cmp < 0:
                self.helper.swap(xs, lo, hi, lt, i)
                lt += 1
                i += 1
            elif cmp > 0:
                self.helper.swap(xs, lo, hi, i, gt)
                gt -= 1
            else:
                i += 1
        return self.Partition(lt, gt)

    @staticmethod
    def swap(xs: List[C], i: int, j: int):
        """
        Used when we don't want to check or count swaps
        """
        xs[i], xs[j] = xs[j], xs[i]

    def get_helper(self) -> Helper[C]:
        return self.helper
